code/datums/controllers/process/mob_ai.dm

```diff
@@ -6,52 +8,71 @@ datum/controller/process/mob_ai
-				if (!M.client) continue
+			var/tickme = FALSE
 
-				if (M.abilityHolder && !M.abilityHolder.composite_owner)
-					if (world.time >= M.abilityHolder.next_update) //after a failure to update (no abbilities!!) wait 10 seconds instead of checking again next process
-						if (M.abilityHolder.updateCounters() > 0)
-							M.abilityHolder.next_update = 1 SECOND
-						else
-							M.abilityHolder.next_update = 10 SECONDS
-				scheck()
+			if ((ticknum % 30) == 0)
+				//call life() with a slowed update rate on mobs we manage that arent part of the standard mobs list
+				if (M.ai?.exclude_from_mobs_list && istype(X, /mob/living))
+					var/mob/living/L = X
+					L.Life(src)
+					scheck()
+				//Lightweight mobs get ticked every 6 seconds
+				if(M.mob_flags & LIGHTWEIGHT_AI_MOB)
+					tickme = TRUE
 
+			//normal mobs get ticked every second, heavyweight mobs get ticked every 0.2 seconds
 			if((M.mob_flags & HEAVYWEIGHT_AI_MOB) || (ticknum % 5) == 0) //either we can tick every time, or we tick every 1 second
+				tickme = TRUE
+
+			if(tickme)
 				var/mob/living/L = M
-				if((isliving(M) && (L.is_npc || L.ai_active) || !isliving(M)))
-					if(istype(X, /mob/living/carbon/human))
-						var/mob/living/carbon/human/H = X
-						if(H.uses_mobai && H.ai)
-							H.ai.tick()
-						else
-							H.ai_process()
+				if(istype(X, /mob/living/carbon/human) && (L.is_npc || L.ai_active))
+					var/mob/living/carbon/human/H = X
+					if(!(H.uses_mobai && H.ai))
+						H.ai_process() //old human AI gets to be special until I get around to removing it
 						scheck()
-					else if(istype(X,/mob/living/critter))
-						var/mob/living/critter/C = X
-						if(C.is_npc && C.ai)
-							C.ai.tick()
-					else if(M.ai)
+						continue
+
+				if(isliving(M) && M.ai)
+					if(!L.is_npc || !M.ai.enabled)
+						continue
+				else if(!M.ai || !M.ai.enabled)
+					continue
+
+				if(M.ai._mobai_being_processed)
+					M.ai._mobai_being_processed++
+					if(M.ai._mobai_being_processed > MOBAI_STUCK_THRESHOLD)
+						logTheThing(LOG_DEBUG, "mobAI process", "!BAD! The mob [constructTarget(M)](\ref[M]) appears to be stuck processing its AI, and has been skipped for [M.ai._mobai_being_processed] ticks. Attempting a reset! You should *really* check why it's slow. AI = [M.ai] AI task = [M.ai?.current_task]")
+						M.ai.current_task?.reset()
+						M.ai.switch_to(M.ai.default_task)
+						M.ai._mobai_being_processed = FALSE
+					if(!ON_COOLDOWN(M, "mobAI_overran_warning", 30 SECONDS))
+						logTheThing(LOG_DEBUG, "mobAI process", "The mob [constructTarget(M)](\ref[M]) overran while processing its AI, and will be skipped for one tick. You should probably check why it's slow. AI = [M.ai] AI task = [M.ai?.current_task]")
+					continue
+
+				SPAWN(0)
+					try
+						M.ai._mobai_being_processed = TRUE
 						M.ai.tick()
-						scheck()
+					catch
+						logTheThing(LOG_DEBUG, "mobAI process", "A runtime was thrown by [constructTarget(M)](\ref[M]) while processing its AI")
```

Comment:
DM try catch does actually expose an exception object that should probably be dumped to the log in some way too (since I think this will prevent it from showing up in the runtime viewer?)